src/main/java/com/google/devtools/build/lib/remote/LoggingInterceptor.java

```diff
@@ -0,0 +1,107 @@
+package com.google.devtools.build.lib.remote;
+
+import static com.google.devtools.build.lib.remote.TracingMetadataUtils.extractRequestMetadata;
+
+import com.google.devtools.build.lib.remote.logging.LoggingHandler;
+import com.google.devtools.build.lib.remote.logging.RpcLogEntry.LogEntry;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import javax.annotation.Nullable;
+
+/** Client interceptor for logging details of certain gRPC calls. */
+public class LoggingInterceptor implements ClientInterceptor {
+
+  /**
+   * Returns a {@link LoggingHandler} to handle logging details for the specific method.
+   *
+   * @param method Method to return handler for.
+   * @return A handler for the given method. Returns null if the method has no specific handler.
+   */
+  protected @Nullable LoggingHandler selectHandler(MethodDescriptor method) {
+    // TODO(cdlee): add handlers for methods
+    return null;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    LoggingHandler<ReqT, RespT> handler = selectHandler(method);
+    if (handler != null) {
+      return new LoggingForwardingCall<>(next.newCall(method, callOptions), handler, method);
+    } else {
+      return next.newCall(method, callOptions);
+    }
+  }
+
+  /**
+   * Wraps client call to log call details by building a {@link LogEntry} and writing it to a log.
+   */
+  private static class LoggingForwardingCall<ReqT, RespT>
+      extends ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT> {
+    private LoggingHandler<ReqT, RespT> handler;
+    private LogEntry.Builder entryBuilder = LogEntry.newBuilder();
```

Comment:
nit: move the full initialization to the constructor.

Reply:
done.